posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,130 @@
+from django.conf import settings
+
+from posthog.clickhouse.table_engines import (
+    Distributed,
+    ReplicationScheme,
+    AggregatingMergeTree,
+)
+
+TABLE_BASE_NAME = "sessions"
+SESSIONS_DATA_TABLE = lambda: f"sharded_{TABLE_BASE_NAME}"
+
+# if updating these column definitions
+# you'll need to update the explicit column definitions in the materialized view creation statement below
+SESSIONS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    -- part of order by so will aggregate correctly
+    session_id VARCHAR,
+    -- part of order by so will aggregate correctly
+    team_id Int64,
+    -- ClickHouse will pick any value of distinct_id for the session
+    -- this is fine since even if the distinct_id changes during a session
+    -- it will still (or should still) map to the same person
+    distinct_id SimpleAggregateFunction(any, String),
+    min_first_timestamp SimpleAggregateFunction(min, DateTime64(6, 'UTC')),
+    max_last_timestamp SimpleAggregateFunction(max, DateTime64(6, 'UTC')),
+
+    urls SimpleAggregateFunction(groupUniqArrayArray, Array(String)),
+    entry_url AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    exit_url AggregateFunction(argMax, String, DateTime64(6, 'UTC')),
+    initial_utm_source AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_campaign AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_medium AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_term AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_content AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_referring_domain AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
```

Comment:
I'd also include
$initial_gclid
$initial_gad_source

as these are used in the channel_type calculation.

Eventually we might have the problem of people wanting other initial properties (fbclid, msclkid, gclsrc, wbraid, gbraid, twclid, li_fat_id, yclid, ttclid, etc). I don't know if we have a good answer for this yet but for now it's worth making a column for the ones we already use

Reply:
the benefit of the properties columns is you don't need to know what you want to store

the problem with the properties columns is that ClickHouse is really slow at querying them... and the bigger the blob the worse that gets

We did a very simple benchmark yesterday and it was like 30x slower to query properties rather than use materialized columns here.

So, _really_ you need to materialize every property that is going to be read into this sessions table... 

And need to consider whether we want to make that automagical for self-hosted too

----

Otherwise this'll be fast for a few weeks and then as the table grows end up being a slow query from a materialized view instead of a slow query from the events table 🫠

Reply:
Cool well here's a full list of what I could figure out from my own knowledge + google + chatgpt

utm_source
utm_medium
utm_campaign
utm_term
utm_content
gclid
gad_source
wbraid
gbraid
fbclid
msclkid
twclid
li_fat_id
yclid
ttclid
dclid
vkclid
pid
igshid
ob_click_id

I'm sure there's more! I don't think we will ever be able to list them all which is why it could be good to have a properties json as a fallback, or alternatively we could just provide initial_param_1 to 5 and let customers define them themselves beyond the major ones.

Reply:
Sorry that's just a list of attribution ids used by major platforms, we don't use all of them, though we do use a few of them.

This is the channel type https://github.com/PostHog/posthog/blob/c94d41ea45890572044d3ffa970f2e983d01da60/posthog/hogql/database/schema/channel_type.py#L34, which is a hogql virtual field, you can do `select $virt_initial_channel_type from persons limit 10` and see it

Reply:
Hmm it might be worth just including the person properties here from PoE, just `any(person.properties)`, so that we get person initial properties